### PR TITLE
feat: add reserve transfer example

### PR DESCRIPTION
**Pull request by kratico**
Add [reserve transfer asset](https://wiki.polkadot.network/docs/learn-xcm-usecases#reserve-asset-transfer) example between Statemine and Trappist.

Notes
- for this example, Statemine acts as the source and reserve chain.
- the example uses [trappist#22755b2](https://github.com/paritytech/trappist/commit/22755b22cb13ff562df50e73872f444066dba3fd) which is compatible with `polkadot v0.9.37` and `polkadot-parachain v0.9.370`

The example steps are
- In Rococo, create a sufficient asset on Statemine (reserve parachain)
  Notes
  - asset should be sufficient to be able to pay fees in Statemine
  - root is needed to create a sufficient asset
  - in a common good parachains, root is accessed through the relay chain
- In Statemine, mint the new reserve asset
- In Trappist, create an asset to pair it with the Statemeine reserve asset 
- In Trappist, register Trappist AssetId to Statemine AssetId
- In Statemine, make a reserve transfer asset to Trappist

### CODE REVIEW

**Comment by kratico** [comment]
@harrysolovay @tjjfvi I need to setup docker to debug the GH action issue
In the meantime, feedback is welcome

**Comment by tjjfvi** [comment]
Let's wait to merge this until after #823

**Comment by tjjfvi** [comment]
@kratico Devnets has been merged, but it'll need to be modified to set up the hrmp channels.

File: devnets/startNetwork.ts
Comment (kratico) [inline devnets/startNetwork.ts:184]: The `trappist` binary does not have the utility to generate the node key.

Ideally, we should figure out the algo to generate the node key and not use a binary to compute it.

File: .trunignore
Comment (tjjfvi) [inline .trunignore:1]: Do these error currently? Or are they just slow?

File: devnets/startNetwork.ts
Comment (tjjfvi) [inline devnets/startNetwork.ts:231]: Can the 8 and 512 constants be extracted? (Should they be configurable?)

File: devnets/startNetwork.ts
Comment (tjjfvi) [inline devnets/startNetwork.ts:227]:
```suggestion
 genesisConfig.hrmp ??= { preopenHrmpChannels: [] }
```

File: devnets/startNetwork.ts
Comment (kratico) [reply devnets/startNetwork.ts:229]: I'll update the function signature and set these values as default.

Given that these values can be different for each permutation, the alternative would be to code each permutation in the `capi.config.ts`

@tjjfvi  WDYT?

File: .trunignore
Comment (harrysolovay) [reply .trunignore]: Let's ignore the XCM examples for now. We can re-enable upon configurability of block times.

File: devnets/startNetwork.ts
Comment (harrysolovay) [reply devnets/startNetwork.ts]: What resulting behavior does this impact? Is it meaningful to JS devs?

Comment (kratico) [reply devnets/startNetwork.ts]: I guess that we could see an impact if there are many messages, but given that this is for dev, I think we should be fine with these values

File: words.txt
Comment (harrysolovay) [inline words.txt:93]:
```suggestion
```

File: examples/xcm/reserve_transfer.eg.ts
Comment (kratico) [inline examples/xcm/reserve_transfer.eg.ts:13]: @harrysolovay for `sudo` calls I switched `alexa` to `alice (or root)` because the default sudo key for these runtimes is `alice` key

**Review by harrysolovay — DISMISSED** [review]
LGTM

File: capi.config.ts
Comment (tjjfvi) [inline capi.config.ts:61]: Can this network be merged with the above once #888 is merged?

File: examples/xcm/reserve_transfer.eg.ts
Comment (tjjfvi) [inline examples/xcm/reserve_transfer.eg.ts:215]: Can you make a codec for this and use that instead?

File: capi.config.ts
Comment (kratico) [reply capi.config.ts:61]: It can't be merged because `trappist` hasn't been updated to XCM v3.
Therefore this example needs a network with `polkadot v0.9.37`, `polkadot v0.9.370` and `trappist#...`

File: examples/xcm/reserve_transfer.eg.ts
Comment (kratico) [reply examples/xcm/reserve_transfer.eg.ts:215]: Would something like `$sovereignAccount.encode(["sibl", 2000])` work?

Comment (kratico) [reply examples/xcm/reserve_transfer.eg.ts:215]: @tjjfvi Where shall I leave this codec?

Comment (harrysolovay) [reply examples/xcm/reserve_transfer.eg.ts:215]: Depends, is this sovereign account constant specific to this setup? If so, `patterns/constants.ts`?

Comment (kratico) [reply examples/xcm/reserve_transfer.eg.ts:215]: Adding more info, parachain sovereign account addresses can be prefixed with the binary representation of `sibl` or `para`

So this would be a codec for parachain addresses.

There are some details on how to compute the address in https://substrate.stackexchange.com/questions/1200/how-to-calculate-sovereignaccount-for-parachain/1210

Comment (harrysolovay) [reply examples/xcm/reserve_transfer.eg.ts:215]: In that case, how about a `patterns/para_id.ts`, which contains two codecs: `$siblId` and `$paraId`?

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:212]: We can probably get rid of these two lines of comments. As well as have `214-217` on a single line.

Comment (harrysolovay) [reply examples/xcm/reserve_transfer.eg.ts:212]: ... unless, do you think it's important to describe this codec usage?

**Review by harrysolovay — DISMISSED** [review]

File: examples/xcm/reserve_transfer.eg.ts
Comment (kratico) [reply examples/xcm/reserve_transfer.eg.ts]: you are right, these comments are not necessary

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:40]: We can use the codegened type factory here.

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:42]: And here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:46]: And here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:118]: And here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:151]: Here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:157]: Here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:166]: Here

File: examples/xcm/reserve_transfer.eg.ts
Comment (harrysolovay) [inline examples/xcm/reserve_transfer.eg.ts:167]: Here

File: devnets/startNetwork.ts
Comment (tjjfvi) [inline devnets/startNetwork.ts:306]: Do hrmp channels need to be opened for both directions? I.e. 1 -> 2 and 2 -> 1?

File: fluent/ExtrinsicRune.ts
Comment (tjjfvi) [inline fluent/ExtrinsicRune.ts:35]:
```suggestion
  callData = this.$call.encoded(this)
```

File: patterns/para_id.ts
Comment (tjjfvi) [inline patterns/para_id.ts:25]:
```suggestion
  const $prefix = $.constant(null, new TextEncoder().encode(prefix))
  const $postfix = $.constant(null, new Uint8Array(24))
  return $.createCodec<number>({
    _metadata: $.metadata(`$${prefix}Id`),
    _staticSize: $prefix._staticSize + $.u32._staticSize + $postfix._staticSize,
    _encode(buffer, value) {
      $prefix._encode(buffer, null)
      $.u32.encode(buffer, value)
      $postfix._encode(buffer, null)
    },
    _decode(buffer) {
      $prefix._decode(buffer)
      const paraId = $.u32._decode(buffer)
      $postfix._decode(buffer)
      return paraId
    },
    _assert(assert) {
      $.u32._assert(assert)
    },
  })
}
```

File: patterns/signature/statemint.ts
Comment (tjjfvi) [inline patterns/signature/statemint.ts:1]: Now that #815 is merged I believe a copy of this file exists already?

File: devnets/startNetwork.ts
Comment (kratico) [reply devnets/startNetwork.ts:306]: yes

File: patterns/signature/statemint.ts
Comment (kratico) [reply patterns/signature/statemint.ts:1]: There is [`westmint.ts`](https://github.com/paritytech/capi/blob/d875ffa0c1a69051d10292a959ac08eaa67fd0ef/patterns/signature/westmint.ts)

Nevertheless, I will update to use `SignatureProps` from `patterns/signature/polkadot.ts`

Comment (tjjfvi) [reply patterns/signature/statemint.ts:1]: Don't westmint and statemine have the same signature types? If so, we only need one file. Probably best named statemine than westmint, though.

Comment (kratico) [inline patterns/signature/statemint.ts:1]: @tjjfvi @harrysolovay `westmint.ts` has been tweaked so it can be used for statemint/statemine too.

What would be a better filename?

Comment (harrysolovay) [reply patterns/signature/statemint.ts:1]: `statemint` should be good 👍

**Review by tjjfvi — APPROVED** [review]

**Pull request merged**
